Computational improvements to static optimization

nathantpickle-cfdrc opened the pull request:
### Brief summary of changes
The static optimization code had inefficiencies due to how the input data were splined. This PR avoids unnecessary splining of data to improve speed of the calculations. The analysis in `analyze1` in `testArm26()` ran approximately 25% faster with these changes.

### Testing I've completed
Ran `testStaticOptimization.cpp` with no failures.

### Looking for feedback on...
Should be ready to merge.

### CHANGELOG.md (choose one)

- no need to update because...this is a minor change for improved speed

nickbianco reviewed (changes requested):
Nice work @nathantpickle-cfdrc! I left a handful of comments, mostly related to formatting (e.g., tabs to spaces), documentation, and one unused variable.

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-:-OLpb-fM9W3nqpdrBy9n:b-1xgcpi)* status: all files reviewed, 8 unresolved discussions (waiting on @nathantpickle-cfdrc)

___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 176 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpYh3158gTeUACvlVL:-OLpYh3158gTeUACvlVM:bba5hcu) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimizationTarget.cpp#L176)):*
> ```cpp
> *
> * @param aStatesDerivativeStore States derivative storage.
> */
> ```

These Doxygen-style comments won't actually appear in Doxygen, but I see that you're copying the existing comments. This is a bit out of scope of this PR, but if you wanted to make another improvement, you could move these comments and the others nearby to the header file so Doxygen can parse them.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 623 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpaJK6EO2xXrYQ8HQy:-OLpaJK6EO2xXrYQ8HQz:bb0g9z) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimizationTarget.cpp#L623)):*
> ```cpp
>         }
> 
> 		// No longer use a function to calcDerivative, but instead use the _statesDerivativeSplineSet from SO FAST 
> ```

Rather than commenting on what we used to do (which developers won't see anymore), just state what is happening now, e.g., "Retrieve state derivatives..."
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 628 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpanmJEaftu2uPqado:-OLpanmJEaftu2uPqadp:bx58miq) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimizationTarget.cpp#L628)):*
> ```cpp
> 		int nu = _model->getNumSpeeds();
> 		Array<double> targetStateArray(0.0, nq + nu);
> 		_statesDerivativeStore->getDataAtTime(t, nq + nu, targetStateArray);
> ```

Is `_statesDerivativeSplineSet` not actually used?
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 629 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpZkEw-3T1oWFVRZU-:-OLpZkEx1HdCbhoMF1Tn:b-umspxw) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimizationTarget.cpp#L629)):*
> ```cpp
> 		Array<double> targetStateArray(0.0, nq + nu);
> 		_statesDerivativeStore->getDataAtTime(t, nq + nu, targetStateArray);
> 		double targetAcceleration = targetStateArray[ind];
> ```

Convert tabs to spaces.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 40 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpYNnX4HY1pGVl2MH6:-OLpYNnX4HY1pGVl2MH7:bzifj0e) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimizationTarget.h#L40)):*
> ```objc
> /**
>  * This class implements static optimization to compute muscle forces and
>  * activations without re-splining during the calculation of state derivatives.
> ```

Re-splining what?
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 71 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpYJnM38WwPXSy1c8N:-OLpYJnM38WwPXSy1c8O:b-6o4haf) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimizationTarget.h#L71)):*
> ```objc
>     GCVSplineSet _statesSplineSet;
> 	const Storage *_statesDerivativeStore;
> 	GCVSplineSet _statesDerivativeSplineSet;
> ```

Remove tabs.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 324 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpXplD4O4Huymtl9zJ:-OLpXplD4O4Huymtl9zK:b22qp00) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimization.cpp#L324)):*
> ```cpp
>     target.setStatesSplineSet(_statesSplineSet);
>     target.setStatesDerivativeStore(_statesDerivativeStore);
> 	target.setStatesDerivativeSplineSet(_statesDerivativeSplineSet);
> ```

Remove tab.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 585 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpXttb4D3MivfNqlot:-OLpXttb4D3MivfNqlou:buwljj4) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimization.cpp#L585)):*
> ```cpp
> 	// Add spline set for the first derivative of states to pass to target for SO FAST and no longer need re-splining to calcDerivative at each time step
> 	_statesDerivativeStore = _statesSplineSet.constructStorage(1);
> 	_statesDerivativeSplineSet = GCVSplineSet(5, _statesDerivativeStore);
> ```

Convert tabs to spaces.

(I actually didn't know about the `constructStorage()` trick for constructing derivatives -- nice!)


<!-- Sent from Reviewable.io -->

nathantpickle-cfdrc reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037)* status: 1 of 4 files reviewed, 8 unresolved discussions (waiting on @nickbianco)

___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 324 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpXplD4O4Huymtl9zJ:-OMNhNlFBTBlb5mbbR6u:b-gjomrr) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimization.cpp#L324)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Remove tab.
</blockquote></details>

This should be fixed throughout the files. 

My linter made a fair number of whitespace changes - please let me know if I need to revert those.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 176 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpYh3158gTeUACvlVL:-OMNhtUb3NK_0YfvPJVA:bgoq6p6) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimizationTarget.cpp#L176)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

These Doxygen-style comments won't actually appear in Doxygen, but I see that you're copying the existing comments. This is a bit out of scope of this PR, but if you wanted to make another improvement, you could move these comments and the others nearby to the header file so Doxygen can parse them.
</blockquote></details>

I moved all the Doxygen-style comments to the header.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 623 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpaJK6EO2xXrYQ8HQy:-OMNhzGD-us5ZrpA-ccK:b-ho27km) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimizationTarget.cpp#L623)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Rather than commenting on what we used to do (which developers won't see anymore), just state what is happening now, e.g., "Retrieve state derivatives..."
</blockquote></details>

Fixed here as well as a couple of other places I noticed.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 628 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpanmJEaftu2uPqado:-OMNi2f04rxFwsXyNCzQ:byduvxz) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimizationTarget.cpp#L628)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Is `_statesDerivativeSplineSet` not actually used?
</blockquote></details>

After taking a second look at this, the `_statesDerivativeSplineSet` was unused and splines of the derivatives are not necessary. It is sufficient for the algorithm to retrieve the state derivatives that were pre-calculated using the splined states.


<!-- Sent from Reviewable.io -->

nickbianco reviewed (changes requested):
@nathantpickle-cfdrc thanks for the nice changes! I left another round of comments.

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-:-OMShq8y6LyfcvXrmylo:beytjxe)* status: all files reviewed, 18 unresolved discussions (waiting on @nathantpickle-cfdrc)

___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 324 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpXplD4O4Huymtl9zJ:-OMShXcK8B0GxG3MLTrU:b-pulwrd) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimization.cpp#L324)):*
<details><summary><i>Previously, nathantpickle-cfdrc wrote…</i></summary><blockquote>

This should be fixed throughout the files. 

My linter made a fair number of whitespace changes - please let me know if I need to revert those.
</blockquote></details>

Since I made some formatting requests already, I went ahead and reviewed the whitespace changes and made a few additional suggestions. However, in general, I would avoid adding a bunch of whitespace changes to an ongoing PR in the future, as it can complicate the review process. 
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 176 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpYh3158gTeUACvlVL:-OMSdc473TZidkmMz9q0:btojbfd) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimizationTarget.cpp#L176)):*
<details><summary><i>Previously, nathantpickle-cfdrc wrote…</i></summary><blockquote>

I moved all the Doxygen-style comments to the header.
</blockquote></details>

Great!
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 314 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMShKWE3iyKjgDN9-KZ:-OMShKWE3iyKjgDN9-K_:b-2ldzpk) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.cpp#L314)):*
> ```cpp
>         const StaticOptimizationTarget* aTarget, double* dx, const Vector& x,
>         Vector& dpdx) {
>     if (aTarget == NULL) return (-1);
> ```

Remove parenthesis from return value.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 318 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMShMbL7T2TsMhcBq81:-OMShMbL7T2TsMhcBq82:b-cgpud3) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.cpp#L318)):*
> ```cpp
>     // CONTROLS
>     int nx = aTarget->getNumParameters();
>     if (nx <= 0) return (-1);
> ```

Remove parenthesis from the return value.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 210 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSewYcD5J_aJplKB-c:-OMSewYcD5J_aJplKB-d:b-nhn7o6) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L210)):*
> ```objc
>     /**
>      * Compute derivatives of performance with respect to the
>      * controls by central differences.  Note that the gradient array should
> ```

minor:



_Suggestion:_
```objc
     * controls by central differences. Note that the gradient array should
```
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 229 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSf7mp2xtO53fHTqEm:-OMSf7mp2xtO53fHTqEn:b45h) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L229)):*
> ```objc
> 
>     /**
>      * Compute performance given parameters.
> ```





_Suggestion:_
```objc
     * Compute the objective function given the optimization parameters.
```
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 232 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSfARU3uKEdcXi9RM5:-OMSfARU3uKEdcXi9RM6:b45h) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L232)):*
> ```objc
>      *
>      * @param parameters Vector of optimization parameters.
>      * @param performance Value of the performance criterion.
> ```





_Suggestion:_
```objc
     * @param performance Value of the objective function.
```
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 239 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSfx_w5eZIlfRiSk-w:-OMSfx_w5eZIlfRiSk-x:undefined) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L239)):*
> ```objc
> 
>     /**
>      * Compute the gradient of performance given parameters.
> ```





_Suggestion:_
```objc
     * Compute the gradient of the objective function given the optimization parameters.
```
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 242 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSg2Fl6zRH6Cr8T5bt:-OMSg2FmBSY6ffZrbADP:undefined) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L242)):*
> ```objc
>      *
>      * @param parameters Vector of optimization parameters.
>      * @param gradient Derivatives of performance with respect to the
> ```





_Suggestion:_
```objc
     * @param gradient Derivatives of the cost function with respect to the
```
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 250 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSg5xh6g-zmCLrFiQz:-OMSg5xh6g-zmCLrFiR-:undefined) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L250)):*
> ```objc
> 
>     /**
>      * Compute acceleration constraints given parameters.
> ```





_Suggestion:_
```objc
     * Compute acceleration constraints given the optimization parameters.
```
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 260 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSgO0Q69w9lzUVmoNA:-OMSgO0Q69w9lzUVmoNB:undefined) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L260)):*
> ```objc
> 
>     /**
>      * Compute the gradient of constraint given parameters.
> ```





_Suggestion:_
```objc
     * Compute the constraint Jacobian given the optimization parameters.
```
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 263 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSgTBnAvfcg6qeV3Xp:-OMSgTBnAvfcg6qeV3Xq:undefined) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L263)):*
> ```objc
>      *
>      * @param parameters Vector of parameters.
>      * @param jac Derivative of constraint with respect to the parameters.
> ```





_Suggestion:_
```objc
     * @param jac The constraint Jacobian with respect to the optimization parameters.
```
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 271 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSgtrT4JTxgAwUyBTC:-OMSgtrT4JTxgAwUyBTD:undefined) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L271)):*
> ```objc
> private:
>     /**
>      * Compute all constraints given parameters.
> ```





_Suggestion:_
```objc
     * Compute all constraints given the optimization parameters.
```
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 249 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSeApa5RAu8l5SCXGs:-OMSeApa5RAu8l5SCXGt:by2agx4) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L249)):*
> ```cpp
>  */
> Storage* StaticOptimization::getActivationStorage() {
>     return (_activationStorage);
> ```

Remove parentheses around the return value.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 331 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSeH0CEQuV262LYzq1:-OMSeH0CEQuV262LYzq2:b-67966e) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L331)):*
> ```cpp
>         // Some IPOPT-specific settings
>         optimizer->setLimitedMemoryHistory(
>                 500); // works well for our small systems
> ```

I would revert this line break. You could delete the comment to fit within the column limit.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 492 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSeVfo5rJu31GXPbSq:-OMSeVfo5rJu31GXPbSr:b-h42qd7) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L492)):*
> ```cpp
>  */
> int StaticOptimization::begin(const SimTK::State& s) {
>     if (!proceed()) return (0);
> ```

Remove parentheses from the return value.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 634 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSea4uD9Mx412yOjQz:-OMSea4uD9Mx412yOjR-:b7qbii3) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L634)):*
> ```cpp
>     record(s);
> 
>     return (0);
> ```

Remove parentheses.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 646 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSebS20qtgv37DhQln:-OMSebS20qtgv37DhQlo:b-2ldzpk) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L646)):*
> ```cpp
>  */
> int StaticOptimization::end(const SimTK::State& s) {
>     if (!proceed()) return (0);
> ```

Remove parenthesis from return value.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 650 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSeeMPE9taTpDIsIxy:-OMSeeMPE9taTpDIsIxz:b-gkpxcc) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L650)):*
> ```cpp
>     record(s);
> 
>     return (0);
> ```

Remove parentheses from return value.


<!-- Sent from Reviewable.io -->

nathantpickle-cfdrc reviewed (commented):
Thanks! In addressing some of the feedback I noticed that the Doxygen comments had gotten stale. I went through and updated the parameter names in the documentation to match the actual parameters.

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-:-ON15gMa-Lfd_oVFvdhU:b-1208de)* status: 1 of 4 files reviewed, 17 unresolved discussions (waiting on @nickbianco)

___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 324 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OLpXplD4O4Huymtl9zJ:-ON10Y3o1rtY8pFz3GVy:bnmcv8t) ([raw file](https://github.com/opensim-org/opensim-core/blob/6e1f95d23c36df1f0767f545f733eb3b70416929/OpenSim/Analyses/StaticOptimization.cpp#L324)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Since I made some formatting requests already, I went ahead and reviewed the whitespace changes and made a few additional suggestions. However, in general, I would avoid adding a bunch of whitespace changes to an ongoing PR in the future, as it can complicate the review process.
</blockquote></details>

Thanks! I'll be careful of that in the future.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 249 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSeApa5RAu8l5SCXGs:-ON10T8p1D5FeNOvpyCr:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L249)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Remove parentheses around the return value.
</blockquote></details>

Done.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 331 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSeH0CEQuV262LYzq1:-ON10cK3DFR4vvW5vt4r:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L331)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

I would revert this line break. You could delete the comment to fit within the column limit.
</blockquote></details>

Done.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 492 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSeVfo5rJu31GXPbSq:-ON10e7x0LVHLiVc1yhw:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L492)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Remove parentheses from the return value.
</blockquote></details>

Done.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 634 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSea4uD9Mx412yOjQz:-ON10iZSBhE8aYLsdAi4:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L634)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Remove parentheses.
</blockquote></details>

Done.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 646 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSebS20qtgv37DhQln:-ON10joBA0WAd4gpgtyZ:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L646)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Remove parenthesis from return value.
</blockquote></details>

Done.
___
*[`OpenSim/Analyses/StaticOptimization.cpp` line 650 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSeeMPE9taTpDIsIxy:-ON10khp52Uq2qbMgffy:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimization.cpp#L650)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Remove parentheses from return value.
</blockquote></details>

Done.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 314 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMShKWE3iyKjgDN9-KZ:-ON11nIX110tBoV1vHjP:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.cpp#L314)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Remove parenthesis from return value.
</blockquote></details>

Done.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 318 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMShMbL7T2TsMhcBq81:-ON11nzICkOnXI--4tsU:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.cpp#L318)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Remove parenthesis from the return value.
</blockquote></details>

Done.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 229 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSf7mp2xtO53fHTqEm:-ON10w-w3I4BWzD4EWfH:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L229)):*
> ```objc
> 
>     /**
>      * Compute performance given parameters.
> ```

Done.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 232 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSfARU3uKEdcXi9RM5:-ON15buJ6BRKl7IP7sUC:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L232)):*
> ```objc
>      *
>      * @param parameters Vector of optimization parameters.
>      * @param performance Value of the performance criterion.
> ```

Done.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 239 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSfx_w5eZIlfRiSk-w:-ON15bFC26ljlvSndguw:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L239)):*
> ```objc
> 
>     /**
>      * Compute the gradient of performance given parameters.
> ```

Done.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 242 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSg2Fl6zRH6Cr8T5bt:-ON11bCh0hsbUH0Z91X4:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L242)):*
> ```objc
>      *
>      * @param parameters Vector of optimization parameters.
>      * @param gradient Derivatives of performance with respect to the
> ```

Done.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 250 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSg5xh6g-zmCLrFiQz:-ON11cpLBJbOlb1Us-VV:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L250)):*
> ```objc
> 
>     /**
>      * Compute acceleration constraints given parameters.
> ```

Done.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 260 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSgO0Q69w9lzUVmoNA:-ON11eQB3FHe0VGSRbZO:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L260)):*
> ```objc
> 
>     /**
>      * Compute the gradient of constraint given parameters.
> ```

Done.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 263 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSgTBnAvfcg6qeV3Xp:-ON11hZ94w6owY2eFSDa:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L263)):*
> ```objc
>      *
>      * @param parameters Vector of parameters.
>      * @param jac Derivative of constraint with respect to the parameters.
> ```

Done.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.h` line 271 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OMSgtrT4JTxgAwUyBTC:-ON11jLA3gsVAZnOECms:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d8a13713ad5ab96155acda3780c140da1f2dfe9c/OpenSim/Analyses/StaticOptimizationTarget.h#L271)):*
> ```objc
> private:
>     /**
>      * Compute all constraints given parameters.
> ```

Done.


<!-- Sent from Reviewable.io -->

nickbianco reviewed (changes requested):
Looks great! Thanks for doing some clean up along the way. 

I have one final request: let's add a change log entry to `CHANGELOG.md`. I think the 25% speed improvement is worth highlighting there. After that, we'll be good to merge.

> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-:-ON17cmr-oOihKbVo-hO:b-6tqnic)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @nathantpickle-cfdrc)



<!-- Sent from Reviewable.io -->

nickbianco commented:
@nathantpickle-cfdrc the style CI test failed. Looks like some tabs snuck back into `StaticOptimization.h`:
```
./OpenSim/Analyses/StaticOptimization.h:82:	Storage *_statesDerivativeStore;
./OpenSim/Analyses/StaticOptimization.h:83:	GCVSplineSet _statesDerivativeSplineSet;
Tabs found in the lines shown above. See CONTRIBUTING.md about tabs.
```

nickbianco reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Thanks @nathantpickle-cfdrc!

> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-:-ON1yWLf1Lby0eOAhVxr:b-eka1l2)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @nathantpickle-cfdrc)



<!-- Sent from Reviewable.io -->

nickbianco reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037)* status: all files reviewed, 1 unresolved discussion (waiting on @nathantpickle-cfdrc)

___
*[`CHANGELOG.md` line 92 at r4](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-ON22LNhBx7ReRT_xo7P:-ON22LNhBx7ReRT_xo7Q:b-4bspvm) ([raw file](https://github.com/opensim-org/opensim-core/blob/d11f28faf6d7b297c3cdcd2a55d11e8487e93738/CHANGELOG.md#L92)):*
> ```Markdown
>   tendon force bounds rather than fiber length bounds. (#4040)
> - Fixed bugs in `PolynomialPathFitter` when too few coordinate samples were provided. (#4039)
> - Pre-calculate state derivatives in static optimization to reduce computational time by as much as 25% (#4037)
> ```

Sorry to add one more comment after giving the thumbs-up, but I think we could add a bit more detail here. Something along the lines of "In `StaticOptimization`, state derivatives are now pre-calculated at the time points of the original data, rather than calculated during optimization from splines. This change reduces computational time by as much as 25%. (#4037)"

I have one last follow up question based on the that: did you double check that the time points the state derivatives are calculated from are consistent with the time points that `StaticOptimization` uses for the optimization? Just double checking that users will get consistent results. (The tolerances in `testStaticOptimization` are a bit loose.)


<!-- Sent from Reviewable.io -->

nickbianco commented:
@nathantpickle-cfdrc, actually, I have one last comment to address. Just want to clarify that these changes won't produce unexpected behavior for users.

nathantpickle-cfdrc reviewed (commented):
No problem - let me run a few more checks to make sure.

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-:-ONMNrYY2wmF8wIK80mt:b-ttdlfz)* status: all files reviewed, 1 unresolved discussion (waiting on @nathantpickle-cfdrc)



<!-- Sent from Reviewable.io -->

nathantpickle-cfdrc reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037)* status: 2 of 6 files reviewed, 3 unresolved discussions (waiting on @nickbianco)

___
*[`CHANGELOG.md` line 92 at r4](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-ON22LNhBx7ReRT_xo7P:-ONpfAY1CgjG-g6920rN:b-xwo0hm) ([raw file](https://github.com/opensim-org/opensim-core/blob/d11f28faf6d7b297c3cdcd2a55d11e8487e93738/CHANGELOG.md#L92)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Sorry to add one more comment after giving the thumbs-up, but I think we could add a bit more detail here. Something along the lines of "In `StaticOptimization`, state derivatives are now pre-calculated at the time points of the original data, rather than calculated during optimization from splines. This change reduces computational time by as much as 25%. (#4037)"

I have one last follow up question based on the that: did you double check that the time points the state derivatives are calculated from are consistent with the time points that `StaticOptimization` uses for the optimization? Just double checking that users will get consistent results. (The tolerances in `testStaticOptimization` are a bit loose.)
</blockquote></details>

@nickbianco - Ryan and I went back and took another look at this. Our understanding is that static optimization iterates through the original timesteps and performs optimization at each timestep without perturbing the time value. If this is correct, it should not affect anything to pre-calculate the state derivatives and store them. 

We added a couple of tests that seem to confirm that the changes will not produce inconsistent results. Let me know your thoughts.
___
*[`Applications/Analyze/test/testStaticOptimization.cpp` line 268 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-ONpesQt0s38vhZD0zpx:-ONpesQu-9fVnPPJGbO5:b-i3tij0) ([raw file](https://github.com/opensim-org/opensim-core/blob/2069bd3d57bfd40a416db340fc7bd43b4c53c259/Applications/Analyze/test/testStaticOptimization.cpp#L268)):*
> ```cpp
>     ASSERT_EQUAL(forces.getColumnLabels().findIndex("TRImed"), -1);
> 
>     StaticOptimization& statOpt =(StaticOptimization&)model.getAnalysisSet().get("StaticOptimization");
> ```

Added a check that the time values of the stored derivatives match the original time values associated with the states.

This check could also be added to a different test if desired.
___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 499 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-ONpd_2i25B34HZ3h3f1:-ONpd_2i25B34HZ3h3f2:b-krfg63) ([raw file](https://github.com/opensim-org/opensim-core/blob/2069bd3d57bfd40a416db340fc7bd43b4c53c259/OpenSim/Analyses/StaticOptimizationTarget.cpp#L499)):*
> ```cpp
>         Array<double> targetStateArray(0.0, nq + nu);
>         
>         //Ensure state time matches stored time associated with derivative
> ```

@nickbianco added this check to confirm that the stored time value matches the time value in the state that is passed in to static opt.

My assumption is that if the optimizer messes with the time in the state, it should trigger this exception. When running `testStaticOptimization.cpp` this exception was not triggered.

If this satisfactorily answers your question, it can probably be removed from the final PR.


<!-- Sent from Reviewable.io -->

nickbianco reviewed (approved):
Thanks @nathantpickle-cfdrc! I think our bases are covered now. Let's remove the block of code in `StaticOptimizationTarget` and then we should be good to merge.

> Reviewed 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-:-OONxuB-AFMy9UHRIn-Q:b-d3o6x7)* status: all files reviewed, 1 unresolved discussion (waiting on @nathantpickle-cfdrc)

___
*[`OpenSim/Analyses/StaticOptimizationTarget.cpp` line 499 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-ONpd_2i25B34HZ3h3f1:-OONxR6n0jXCZwcxd_vU:b1foa4w) ([raw file](https://github.com/opensim-org/opensim-core/blob/2069bd3d57bfd40a416db340fc7bd43b4c53c259/OpenSim/Analyses/StaticOptimizationTarget.cpp#L499)):*
<details><summary><i>Previously, nathantpickle-cfdrc wrote…</i></summary><blockquote>

@nickbianco added this check to confirm that the stored time value matches the time value in the state that is passed in to static opt.

My assumption is that if the optimizer messes with the time in the state, it should trigger this exception. When running `testStaticOptimization.cpp` this exception was not triggered.

If this satisfactorily answers your question, it can probably be removed from the final PR.
</blockquote></details>

I think your test and the existing tests are satisfactory to address my comment about preserving the time vector. Let's remove this block of code to avoid the overhead of `findIndex()` within the static optimization target.


<!-- Sent from Reviewable.io -->

nickbianco reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037)* status: all files reviewed, 1 unresolved discussion (waiting on @nathantpickle-cfdrc)



<!-- Sent from Reviewable.io -->

nathantpickle-cfdrc commented:
Will do. We also had to create a `getStatesDerivativeStore()` method in `StaticOptimization.h` to allow us to retrieve the stored state derivatives for the test we added to `testArm26DisabledMuscles()`. That method is not likely to be used anywhere besides the test. What do you think about removing both `getStatesDerivativeStore()` and the test we added?





<!-- Sent from Reviewable.io -->

nickbianco reviewed (changes requested):
I think we should keep the test, and I think it's fine to keep `getStatesDerivativeStore()`.

I added one quick request for the added test. After addressing that, we should be good to merge. 

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-:-OOcJmD15jhlNb2kgyfe:b-h2rx79)* status: all files reviewed, 2 unresolved discussions (waiting on @nathantpickle-cfdrc)

___
*[`Applications/Analyze/test/testStaticOptimization.cpp` line 268 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OOcJdL92s5UrunWqJGH:-OOcJdL92s5UrunWqJGI:b-1sjr9u) ([raw file](https://github.com/opensim-org/opensim-core/blob/2069bd3d57bfd40a416db340fc7bd43b4c53c259/Applications/Analyze/test/testStaticOptimization.cpp#L268)):*
> ```cpp
>     ASSERT_EQUAL(forces.getColumnLabels().findIndex("TRImed"), -1);
> 
>     StaticOptimization& statOpt =(StaticOptimization&)model.getAnalysisSet().get("StaticOptimization");
> ```

Add a brief comment describing this this test.


<!-- Sent from Reviewable.io -->

nathantpickle-cfdrc reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037)* status: 4 of 6 files reviewed, 1 unresolved discussion (waiting on @nickbianco)

___
*[`Applications/Analyze/test/testStaticOptimization.cpp` line 268 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-OOcJdL92s5UrunWqJGH:-OOcP8KvDZ0pdJXozRM_:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/2069bd3d57bfd40a416db340fc7bd43b4c53c259/Applications/Analyze/test/testStaticOptimization.cpp#L268)):*
<details><summary><i>Previously, nickbianco (Nick Bianco) wrote…</i></summary><blockquote>

Add a brief comment describing this this test.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

nickbianco commented:
Thanks @nathantpickle-cfdrc, the code is looking good now. 

Sorry to keep moving the finish line on you, but I'm going to ask you for one more sanity check: could you plot the results from the static optimization tests before and after this change? I think it would be helpful to confirm that everything works qualitatively, especially since the regression tests have fairly loose tolerances.

nathantpickle-cfdrc reviewed (commented):
No problem - we don't want to inadvertently break anything! Ryan helped out and we ran a couple more tests.

We used the OpenSim 4.3 GUI to test the original static optimization code, and ran static optimization from the command line using the same setup file to test the modified code.

The first test we ran was a torque-actuated double pendulum. There are two torque actuators, tau0 and tau1. The plots are exactly overlaid, so the blue "original" line is not visible.

![pendulum.png](https://files.reviewableusercontent.io/images/reviewable/119360930/CspmfszhAuzqu6o8XP50Y3og/pendulum.png)

We also ran a more complex model to provide a more rigorous stress test of the updated algorithm. For that, we used a 5-second simulation of the Vasavada neck model performing a motion from 30 degrees to 60 degrees of head yaw. We plotted several of the muscle results and reserve actuators and found that they also overlaid each other exactly.

![scalenus\_med\_r.png](https://files.reviewableusercontent.io/images/reviewable/119360930/ATz2s7xoVBUtld32rVAbHnBW/scalenus_med_r.png)
![iliocost\_cerv\_c5rib\_l.png](https://files.reviewableusercontent.io/images/reviewable/119360930/Qv_kSl7p0gKYXHbVPCc_BVld/iliocost_cerv_c5rib_l.png)
![c2\_c1\_ry\_reserve.png](https://files.reviewableusercontent.io/images/reviewable/119360930/Jbf9fmdfD7r8EQaNbigqIcb0/c2_c1_ry_reserve.png)
![deepmult-t2-t1\_l.png](https://files.reviewableusercontent.io/images/reviewable/119360930/QtKg3mi4syTLj48AeSFC8jGD/deepmult-t2-t1_l.png)

Happy to run any other test cases you'd like to see before completing this merge.

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/4037#-:-OOhRZUJ7PT8PyumGHB4:b-x010wd)* status: 4 of 6 files reviewed, 1 unresolved discussion (waiting on @nickbianco)



<!-- Sent from Reviewable.io -->

nickbianco commented:
Thanks @nathantpickle-cfdrc! Looks great. I think we've officially covered all the bases. 

Merging!
